File: FungeeCommandsManager/src/main/java/org/openredstone/commands/Derp.java
package org.openredstone.commands;

import net.md_5.bungee.api.ChatColor;
import net.md_5.bungee.api.CommandSender;
import net.md_5.bungee.api.ProxyServer;
import net.md_5.bungee.api.chat.ComponentBuilder;
import net.md_5.bungee.api.chat.TextComponent;
import org.openredstone.handlers.DerpHandler;

public class Derp extends FunCommand {
    private final ProxyServer proxy;
    private final DerpHandler handler;

    public Derp(ProxyServer proxy, DerpHandler handler) {
        super("derp");
        this.proxy = proxy;
        this.handler = handler;
    }

    @Override
    public void execute(CommandSender sender, String[] args) {
        if (args.length > 1) {
            sendUsage(sender, "too many arguments");
            return;
        }

        if (args.length == 0) {
            sendDerp(sender, handler.randomDerp());
        } else {
            sendDerpByIndex(sender, args[0]);
        }
    }

    private void sendDerp(CommandSender sender, String derp) {
        proxy.broadcast(buildDerp(
                sender.getName(),
                derp
        ));
    }

    private void sendDerpByIndex(CommandSender sender, String index) {
        // this is sorta hacky
        try {
            int derpIndex = Integer.valueOf(index);
            sendDerp(sender, handler.derpByIndex(derpIndex));
        } catch (NumberFormatException | IndexOutOfBoundsException e) {
            sendUsage(
                    sender,
                    "argument must be a 0-based number within range"
            );
        }
    }

    private void sendUsage(CommandSender sender, String message) {
        // kinda weird, and not very user-friendly error messages
        sendErrors(
                sender,
                message,
                "/derp derpNumber",
                "random derp: /derp",
                "There are currently " + handler.derpCount() + " derps"
        );
    }

    private void sendErrors(CommandSender sender, String... messages) {
        for (String message : messages) {
            sender.sendMessage(new ComponentBuilder(message).color(ChatColor.RED).create());
        }
    }

    private TextComponent buildDerp(String playerName, String derp) {
        return new TextComponent(
                new ComponentBuilder(" * ").color(ChatColor.GREEN)
                        .append(playerName).color(ChatColor.WHITE)
                        .append(" DERP! ").color(ChatColor.DARK_BLUE)
                        .append(derp).color(ChatColor.LIGHT_PURPLE)
                        .create()
        );
    }
}

File: fungee-manager/src/main/kotlin/org/openredstone/fungee/manager/commands/Derp.kt
package org.openredstone.fungee.manager.commands

import net.md_5.bungee.api.ChatColor
import net.md_5.bungee.api.CommandSender
import net.md_5.bungee.api.ProxyServer
import net.md_5.bungee.api.chat.ComponentBuilder
import net.md_5.bungee.api.chat.TextComponent
import net.md_5.bungee.api.plugin.Command
import org.openredstone.fungee.manager.handlers.DerpHandler

class Derp(private val proxy: ProxyServer, private val handler: DerpHandler) : Command("derp", "funcommands.derp") {
    override fun execute(
        sender: CommandSender,
        args: Array<String>
    ) = when {
        args.size > 1 -> {
            sender.sendUsage("too many arguments")
        }
        args.isEmpty() -> {
            sender.sendDerp(handler.derps.random())
        }
        else -> {
            sender.sendDerpByIndex(args[0])
        }
    }

    private fun CommandSender.sendDerp(derp: String) {
        proxy.broadcast(buildDerp(name, derp))
    }

    private fun CommandSender.sendDerpByIndex(index: String) {
        index
            .toIntOrNull()
            ?.let(handler.derps::getOrNull)
            ?.let { sendDerp(it) }
            ?: sendUsage("argument must be a 0-based number within range")
    }

    private fun CommandSender.sendUsage(message: String) {
        // kinda weird, and not very user-friendly error messages
        arrayOf(
            message,
            "/derp derpNumber",
            "random derp: /derp",
            "There are currently " + handler.derps.size + " derps"
        ).forEach { sendError(it) }
    }

    private fun CommandSender.sendError(message: String) = sendMessage(
        *ComponentBuilder(message).color(ChatColor.RED).create()
    )

    private fun buildDerp(playerName: String, derp: String) = TextComponent(
        *ComponentBuilder(" * ").color(ChatColor.GREEN)
            .append(playerName).color(ChatColor.WHITE)
            .append(" DERP! ").color(ChatColor.DARK_BLUE)
            .append(derp).color(ChatColor.LIGHT_PURPLE)
            .create()
    )

}
